feat: pathshorten() optional trim length

talmobi opened the pull request:
##### What
Second optional argument for pathshorten() to specify preferred trimming length.

##### Example:
`echo pathshorten("/Users/mollie/Documents/some-file.txt", 3)`

###### output 
`/Use/mol/Doc/some-file.txt`

##### Why
Sometimes pathshorten is too short.

For example when pathshortening filenames in the qf/loc list (using the qf/loc entry "module" field added in __v8.0.1782__). The single character shortenings is a bit too much. But limiting it to maybe 3-5 chars would be really nice.

##### Changes
* tests added to `/src/testdir/test_functions.vim` within existing `func Test_pathshorten()` block.
* shorten_dir2 fn added (`src/filepath.c`)
* shorten_dir refactored to use shorten_dir2 with optional trim_len variable set to 1 by default resulting in the same behaviour as before.
*  `src/proto/filepath.pro` and `src/evalfunc.c` updated.

#### How to build and test

##### from scratch
```bash
git clone --branch talmobi/feat/pathshorten-trim-size https://github.com/talmobi/vim
cd vim
./configure
make
cd src
make test_functions
```

##### existing clone
```bash
git checkout -b talmobi/feat/pathshorten-trim-size
git pull https://github.com/talmobi/vim talmobi/feat/pathshorten-trim-size
make distclean
./configure
make
cd src
make test_functions
```

talmobi committed:
test: pathshorten with optional argument for preferred trimming size

talmobi committed:
feat: configure optional second variable to pathshorten()

talmobi committed:
feat: handle optional second trim_len variable using shorten_dir2

talmobi committed:
refactor: use shortend_dir2 from within shorten_dir

talmobi committed:
feat: refactor shorten_dir code into shorten_dir2 and add second optional variable for preferred dirname chunk/trim length

talmobi committed:
feat: add shorten_dir2

brammool commented:
This can be useful.
Please also update the help text.
Please adjust to match the Vim code style.

talmobi committed:
feat: update help text

talmobi committed:
test: cover min shortening length of 1

talmobi committed:
style: use int instead of size_t for optional trim_len second parameter for pathshorten()

problem: size_t usually denotes sizes of objects, trim_len does not fit
this description as it's just the preferred length in characters of the
dir name component. Usually in ranges of 1-10 makes sense.

ref: https://www.quora.com/Why-do-some-C-programs-use-size_t-instead-of-int-What-are-the-advantages

solution: use int instead of size_t

talmobi commented:
Hello there,

I updated the help text.

I switched to use int instead of size_t. Although I'm not sure that's what you meant by vim code style. I tried to ape after existing code.

I'm not sure why some of the CI builds fail. It seems unrelated to this particular PR.

Please let me know what else needs fixing~

dpelle commented on src/filepath.c:
```diff
@@ -1358,17 +1358,26 @@ f_mkdir(typval_T *argvars, typval_T *rettv)
+	trim_len = (int)tv_get_number(&argvars[1]);
+	if (trim_len < 1) trim_len = 1;
+    } else {
+	trim_len = (int)1;
```
The cast it useless.

talmobi replied:
for both? or only L1367?

I mean yeah I guess L1367 is definitely useless ^^;;

dpelle commented on src/filepath.c:
```diff
@@ -1358,17 +1358,26 @@ f_mkdir(typval_T *argvars, typval_T *rettv)
+    if (argvars[1].v_type != VAR_UNKNOWN) {
+	trim_len = (int)tv_get_number(&argvars[1]);
+	if (trim_len < 1) trim_len = 1;
+    } else {
```
In Vim code:
*  curly braces go on a separate line.
* Indentation should be 4 spaces

dpelle commented on src/filepath.c:
```diff
@@ -1358,17 +1358,26 @@ f_mkdir(typval_T *argvars, typval_T *rettv)
+
+    if (argvars[1].v_type != VAR_UNKNOWN) {
+	trim_len = (int)tv_get_number(&argvars[1]);
+	if (trim_len < 1) trim_len = 1;
```
In Vim style, this line should be split  i.e. `trim_len = 1;` on separate line.

talmobi committed:
style: fixed code style errors pointed out by @dpelle

changes:
        removed useless (int) casting
        curly braces on newlines
        else on newlines
        if on newlines
        4 space indentation

talmobi commented:
@dpelle thank you, much appreciated~

talmobi commented:
great success~
